src/agent/src/initdata.rs

```diff
@@ -0,0 +1,639 @@
+};
+
+use anyhow::{anyhow, bail, Context, Result};
+use base64::Engine;
+use const_format::concatcp;
+use serde::Deserialize;
+use sha2::{Digest, Sha256, Sha384, Sha512};
+use slog::Logger;
+use tokio::{fs, process::Command};
+
+use crate::{cdh::CDHClient, config::GuestComponentsProcs, AGENT_CONFIG};
+
+/// Convenience macro to obtain the scope logger
+macro_rules! sl {
+    () => {
+        slog_scope::logger()
+    };
+}
+
+/// Unix socket prefix
+const UNIX_SOCKET_PREFIX: &str = "unix://";
+
+/// Now only initdata `0.1.0` is defined.
+const DEFAULT_INITDATA_VERSION: &str = "0.1.0";
+
+/// Timeout to launch guest components
+const DEFAULT_LAUNCH_PROCESS_TIMEOUT: i32 = 6;
+
+/// Path of AA binary
+const AA_PATH: &str = "/usr/local/bin/attestation-agent";
+
+/// Path to AA unix socket.
+const AA_ATTESTATION_SOCKET: &str =
+    "/run/confidential-containers/attestation-agent/attestation-agent.sock";
+
+/// Path of AA config, which is writable inside TEE.
+const AA_CONFIG_PATH: &str = "/run/aa.toml";
+
+/// Binary path of CDH
+const CDH_PATH: &str = "/usr/local/bin/confidential-data-hub";
+
+/// Path to CDH unix socket.
+const CDH_SOCKET: &str = "/run/confidential-containers/cdh.sock";
+
+/// Unix socket of CDH
+pub const CDH_SOCKET_URI: &str = concatcp!(UNIX_SOCKET_PREFIX, CDH_SOCKET);
+
+/// Path of CDH config, which is writable inside TEE.
+const CDH_CONFIG_PATH: &str = "/run/cdh.toml";
+
+/// Path of ocicrypt config file. This is used by image-rs when decrypting image.
+const OCICRYPT_CONFIG_PATH: &str = "/tmp/ocicrypt_config.json";
+
+/// Path to ASR binary
+const API_SERVER_PATH: &str = "/usr/local/bin/api-server-rest";
+
+/// If no initdata is given, a default (all zero) initdata will be used.
+const DEFAULT_INITDATA_DIGEST: &str = "AAAAAAAAAAAA";
+
+/// Initdata defined in
+/// <https://github.com/confidential-containers/trustee/blob/47d7a2338e0be76308ac19be5c0c172c592780aa/kbs/docs/initdata.md>
+#[derive(Deserialize)]
+pub struct Initdata {
+    version: String,
+    algorithm: String,
+    data: DefinedFields,
+}
+
+static ALREADY_SET_INITDATA: AtomicBool = AtomicBool::new(false);
+
+/// Well-defined keys for initdata of kata/CoCo
+#[derive(Deserialize, Default)]
+#[serde(deny_unknown_fields)]
+pub struct DefinedFields {
+    #[serde(rename = "aa.toml")]
+    aa_config: Option<String>,
+    #[serde(rename = "cdh.toml")]
+    cdh_config: Option<String>,
+    #[serde(rename = "policy.rego")]
+    policy: Option<String>,
+}
+
+pub async fn do_set_initdata(req: &protocols::agent::SetInitdataRequest) -> Result<()> {
+    debug!(sl!(), "SetInitdata called.");
+    // ensure the function could be called only once.
+    if ALREADY_SET_INITDATA
+        .compare_exchange(false, true, Ordering::SeqCst, Ordering::SeqCst)
+        .is_ok()
+    {
+        debug!(sl!(), "SetInitdata called the first time.");
+    } else {
+        bail!("SetInitdata already called");
+    }
+
+    let (initdata_digest, configs) = match req.initdata.is_empty() {
```

Comment:
nit: if blocks can return values. it's more idiomatic and saves a level of indentation

```suggestion
    let (initdata_digest, configs) = if req.initdata.is_empty() {
      ...
    else {
      ...
    };  
```

Reply:
Good. I often thought that `match` would make the logic more clear. It's fine to me to change this into `if`.